Safe liveness checking module and guard (#7540)
ethereum-optimism/optimism: specs/safe-liveness-checking.md

Reviewer: Thank you for upholding our engineering standard and writing this spec about the implementation!

are otherwise inactive. It is not intended to be used to remove signers who are acting in bad faith,
or any other subjective criteria, such cases should be addressed by governance, and the removal
handled via the standard Safe ownership management functionality.

## Liveness checking methodology

This is achieved using two types of contracts which the Safe contract has built-in support for:

1. **Guard contracts:** can execute pre- and post- transaction checks.
1. **Module contracts:** a contract which is added to the Safe by the signers, and thenceforth is
   authorized to execute transactions via the Safe. This means the module must properly implement
   auth conditions internally.

### The Liveness Guard

For implementing liveness checks a `LivenessGuard` is created which receives the signatures from
each executed transaction, and tracks the latest time at which a transaction was signed by each
signer. This time is made publicly available by calling a `lastLive(address)(Timestamp)` method.

Signers may also call the contract's `showLiveness()()` method directly in order to prove liveness.

### The Liveness Module

A `LivenessModule` is also created which does the following:

> 1. Has a function `removeOwner()` that anyone may call to specify an owner to be removed from the

Reviewer 2: ```suggestion
1. Has a function `removeOwners()` that anyone may call to specify one or more owners to be removed from the
```

   Safe.
1. The Module would then check the `LivenessGuard.lastLive()` to determine if the signer is
   eligible for removal.
1. If so, it will call the Safe's `removeSigner()` to remove the non-live signer, and if necessary
   reduce the threshold.
1. When a member is removed, the signing parameters are modified such that `M/N` is the lowest ratio
   which remains greater than or equal to 75%. Using integer math, this can be expressed as `M = (N * 75 + 99) / 100`.

### Owner removal call flow

The following diagram illustrates the flow for removing a single owner. The `verifyFinalState`
box indicates calls to the Safe which ensure the final state is valid.

> ```mermaid

Reviewer 2: This is a great diagram

sequenceDiagram
    participant User
    participant LivenessModule
    participant LivenessGuard
    participant Safe
    User->>LivenessModule: removeOwners([previousOwner], [owner])
    LivenessModule->>LivenessGuard: lastLive(owner)
    LivenessModule->>Safe: getOwners()
    LivenessModule->>Safe: removeOwner(previousOwner, owner)

    alt verifyFinalState
    LivenessModule->>Safe: getOwners()
    LivenessModule->>Safe: getThreshold()
    LivenessModule->>Safe: getGuard()
    end
```

### Shutdown

In the unlikely event that the signer set (`N`) is reduced below the allowed threshold, then (and only then) is a
   shutdown mechanism activated which removes the existing signers, and hands control of the
   multisig over to a predetermined entity.

### Security Properties

> The following security properties must be upheld:

Reviewer 2: Additional property: It must be impossible for the guard's `checkTransaction` or `checkAfterExecution` to permanently revert given any calldata and the current state.

1. Signatures are assigned to the correct signer.
1. Non-signers are unable to create a record of having signed.

> 1. A signer cannot be censored or grieffed such that their signing is not recorded.

Reviewer 2: ```suggestion
1. A signer cannot be censored or griefed such that their signing is not recorded.
```

1. Signers may demonstrate liveness either by signing a transaction or by calling directly to the
   guard.
1. The module only removes a signer if they have demonstrated liveness during the interval, or
     if necessary to convert the safe to a 1 of 1.
1. The module sets the correct 75% threshold upon removing a signer.
1. During a shutdown the module correctly removes all signers, and converts the safe to a 1 of 1.

### Interdependency between the Guard and Module

The Guard has no dependency on the Module, and can be used independently to track liveness of
Safe owners. The Module however does have a dependency on the Guard, only one guard contract can
be set on the Safe, and the Module will be unable to function if the Guard is removed.

> ### Deployment

Reviewer 2: Thanks, this is helpful. Can we add similar steps for removing and/or updating the guard and/or module?

Author: https://github.com/ethereum-optimism/optimism/commit/7cb57c050420d409fb54541cdad348181d57853f

> The module are guard are intended to be deployed and installed on the safe in the following sequence:

Reviewer 2: ```suggestion
The module and guard are intended to be deployed and installed on the safe in the following sequence:
```

# Safe Liveness Checking

<!-- START doctoc generated TOC please keep comment here to allow auto update -->

> <!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

Reviewer 2: TOC is out of date and needs to be regenerated

**Table of Contents**

- [Overview](#overview)
- [Liveness Checking Mechanism](#liveness-checking-mechanism)
- [Liveness Check Implementation](#liveness-check-implementation)
- [Liveness Check Validation](#liveness-check-validation)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Liveness Checking Mechanism

> The Security Security Council uses a specially extended multisig...

Reviewer 2: Looks like this sentence is incomplete

## Liveness checking methodology

This is achieved using two types of contracts which the Safe contract has built-in support for:

1. **Guard contracts:** can execute pre- and post- transaction checks.
1. **Module contracts:** a contract which is added to the Safe by the signers, and thenceforth is
   authorized to execute transactions via the Safe. This means the module must properly implement
   auth conditions internally.

### The Liveness Guard

For implementing liveness checks a `LivenessGuard` is created which receives the signatures from
each executed transaction, and tracks the latest time at which a transaction was signed by each
signer. This time is made publicly available by calling a `lastSigned(address)(Timestamp)` method.

> Signers may also call the contract directly in order to prove liveness.

Reviewer 2: So the `LivenessGuard` only works for executed transactions, but transactions that are signed and don't meet the execution threshold get ignored, since they never land on-chain. This is arguably ok since if a signer is genuinely affected by this, they can just call the contract to prove liveness.

Though, a disinterested signer can repeatedly call `showLiveness` to stay live without actually contributing anything. Maybe worth seeing if all signatures can be managed on-chain instead of relying on Safe's servers to store sigs? Then you can remove the `showLiveness` method and only rely on signatures

Author: Good points.

Regarding unexecuted but signed tx, I agree that's annoying for signers, but should be sufficiently mitigated by `showLiveness`. 

WRT a 'disinterested signer' (I assume you mean one that's not actively participating), I think it's OK. We should keep this guard/module pure in that they only enforce liveness checking. Active participation is a more subjective consideration, and it should be up to governance vote for, and the other signers to enact, the removal of an inactive signer.

> Signers may also call the contract directly in order to prove liveness.

Reviewer 2: Nit, might be nice to include the method name here with the same syntax as `lastSigned(address)(Timestamp)` above (or ignore the parentheses if there's no args, not sure what the usual syntax used in this repo is)
```suggestion
Signers may also call the contract's `showLiveness()()` method directly in order to prove liveness.
```

Reviewer 3: There is a tradeoff with baking in the ABI, it can go stale. Previously a lot of the specs went stale in certain ways but this doesn't seem like it is something that will go stale

Reviewer 3: It looks like other places in this file it uses functions so using function names is fine

Author: Yes, I don't think this spec will evolve too quickly/often once finalized, so it should be OK to include function names.

### The Liveness Module

A `LivenessModule` is also created which does the following:

> 1. Has a function `removeSigner()` that anyone may call to specify a signer to be removed from the

Reviewer 2: The contract calls them owners, the spec calls them signers, probably should align on one of those
```suggestion
1. Has a function `removeOwner()` that anyone may call to specify an owner to be removed from the
```

   Safe.
1. The Module would then check the `LivenessGuard.lastSigned()` to determine if the signer is
   eligible for removal.
1. If so, it will call the Safe's `removeSigner()` to remove the non-live signer, and if necessary
   reduce the threshold.
1. When a member is removed, the signing parameters are modified such that `M/N` is the lowest ratio
   which remains above 75%. These ratios are (9 of 12, 9 of 11, 8 of 10, 7 of 9, 6 of 8). Using
   integer math, this can be expressed as `M = (N * 75 + 99) / 100`.

### Shutdown

In the unlikely event that the signer set (`N`) is reduced below 8, then (and only then) is a
   shutdown mechanism activated which removes the existing signers, and hands control of the
   multisig over to a predetermined entity.

> ### Security Properties

Reviewer 2: Is this section supposed to be prescriptive, i.e. "here's the security properties we must fulfill", or descriptive, i.e. "given the above spec, here's the resulting security properties"?

Author: ```suggestion
### Security Properties

The following security properties must be upheld: 
```

1. Signatures are assigned to the correct signer.
2. Non-signers are unable to create a record of having signed.

> 3. A signer cannot be censored or grieffed such that their signing is not recorded.

Reviewer 2: ```suggestion
3. A signer cannot be censored or griefed such that their signing is not recorded.
```

4. Signers may demonstrate liveness either by signing a transaction or by calling directly to the
   guard.

> 5. The module implements the correct checks priort to removing a signer.

Reviewer 2: ```suggestion
5. The module implements the correct checks prior to removing a signer.
```

1. **Module contracts:** a contract which is added to the Safe by the signers, and thenceforth is
   authorized to execute transactions via the Safe. This means the module must properly implement
   auth conditions internally.

### The Liveness Guard

For implementing liveness checks a `LivenessGuard` is created which receives the signatures from
each executed transaction, and tracks the latest time at which a transaction was signed by each
signer. This time is made publicly available by calling a `lastLive(address)(Timestamp)` method.

Signers may also call the contract directly in order to prove liveness.

### The Liveness Module

A `LivenessModule` is also created which does the following:

1. Has a function `removeOwner()` that anyone may call to specify an owner to be removed from the
   Safe.
1. The Module would then check the `LivenessGuard.lastLive()` to determine if the signer is
   eligible for removal.
1. If so, it will call the Safe's `removeSigner()` to remove the non-live signer, and if necessary
   reduce the threshold.
1. When a member is removed, the signing parameters are modified such that `M/N` is the lowest ratio

>    which remains above 75%. Using integer math, this can be expressed as `M = (N * 75 + 99) / 100`.

Reviewer 2: Clarifying that 75% is allowed
```suggestion
   which remains greater than or equal to 75%. Using integer math, this can be expressed as `M = (N * 75 + 99) / 100`.
```

# Safe Liveness Checking

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Liveness Checking Mechanism](#liveness-checking-mechanism)
- [Liveness checking methodology](#liveness-checking-methodology)
  - [The Liveness Guard](#the-liveness-guard)
  - [The Liveness Module](#the-liveness-module)
  - [Shutdown](#shutdown)
  - [Security Properties](#security-properties)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

> 

Reviewer 3: A mermaid diagram could be very helpful here showing the call flow, chatgpt is good at generating them if you just tell it what to do and iterate on it through multiple prompts

Author: done

1. **Guard contracts:** can execute pre- and post- transaction checks.
1. **Module contracts:** a contract which is added to the Safe by the signers, and thenceforth is
   authorized to execute transactions via the Safe. This means the module must properly implement
   auth conditions internally.

### The Liveness Guard

For implementing liveness checks a `LivenessGuard` is created which receives the signatures from
each executed transaction, and tracks the latest time at which a transaction was signed by each
signer. This time is made publicly available by calling a `lastSigned(address)(Timestamp)` method.

Signers may also call the contract directly in order to prove liveness.

### The Liveness Module

A `LivenessModule` is also created which does the following:

1. Has a function `removeOwner()` that anyone may call to specify an owner to be removed from the
   Safe.
1. The Module would then check the `LivenessGuard.lastSigned()` to determine if the signer is
   eligible for removal.
1. If so, it will call the Safe's `removeSigner()` to remove the non-live signer, and if necessary
   reduce the threshold.

> 1. When a member is removed, the signing parameters are modified such that `M/N` is the lowest ratio

Reviewer 3: This implies that there is a min value for `minOwners` which is not the case

   which remains above 75%. These ratios are (9 of 12, 9 of 11, 8 of 10, 7 of 9, 6 of 8). Using
   integer math, this can be expressed as `M = (N * 75 + 99) / 100`.

### Shutdown

> In the unlikely event that the signer set (`N`) is reduced below 8, then (and only then) is a

Reviewer 3: I think we should define some constants in a box at the top of the file, such as `MIN_SIGNERS: 8` and then use the name of the variable here

Reviewer 3: This could also include liveliness interval as the mainnet value and the fallback address as "optimism foundation address"

Author: I'm not sure about that, are there other examples of that in the specs?
Do we need to specify values for mainnet and tesnet?

Author: Going to pass on this in favor of consistency with other specs docs. Feel free to reopen.

   shutdown mechanism activated which removes the existing signers, and hands control of the
   multisig over to a predetermined entity.

### Security Properties

The following security properties must be upheld:

1. Signatures are assigned to the correct signer.
2. Non-signers are unable to create a record of having signed.
3. A signer cannot be censored or griefed such that their signing is not recorded.
4. Signers may demonstrate liveness either by signing a transaction or by calling directly to the
   guard.

> 5. The module implements the correct checks prior to removing a signer.

Reviewer 3: What are the correct checks?

Author: Added:
> 1. The module only removes a signer if they have demonstrated liveness during the interval, or
     if necessary to convert the safe to a 1 of 1.

> 6. The module sets the correct threshold upon removing a signer.

Reviewer 3: What is the correct threshold? Could link to another section in this doc

Author: I'd rather not commit to a specific value in this doc.
